File: RxEpics/python/examples/resilient_monitor.py
"""The end-to-end resilience demo: values, per-update errors, and link state
as one merged stream — errors and status transitions are messages, not
exceptions that stop the process.

Run this, then in another shell restart the IOC out from under it:

    docker compose stop epics-ioc
    docker compose start epics-ioc

The stream survives: link goes DOWN, no traceback, and once the IOC comes
back the link goes UP and values resume — caproto re-arms the CA
subscription on its own, with no client-side action.

Usage:
    python resilient_monitor.py <pv_name>

Example:
    python resilient_monitor.py TEST:CALC
"""

import asyncio
import sys
import time
from pathlib import Path

sys.path.insert(0, str(Path(__file__).parent.parent / "src"))

import reactivex as rx
from reactivex.scheduler.eventloop import AsyncIOScheduler
from caproto.asyncio.client import Context

from rxepics.connection import connection_status
from rxepics.monitor import monitor_errors, monitor_pv


async def main():
    if len(sys.argv) < 2:
        print("Usage: resilient_monitor.py <pv_name>", file=sys.stderr)
        sys.exit(1)

    pv_name = sys.argv[1]

    loop = asyncio.get_running_loop()
    scheduler = AsyncIOScheduler(loop)
    ctx = Context()

    def ts():
        return int(time.time() * 1000)

    # Three independent streams, tagged and merged into one console view.
    # Each is a message on the wire, never a terminal notification — a bad
    # update or a dropped link shows up as a line of output, not a crash.
    values = monitor_pv(pv_name, ctx).pipe(rx.operators.map(lambda v: f"[{ts()}]  value      {v}"))
    errors = monitor_errors(pv_name, ctx).pipe(rx.operators.map(lambda e: f"[{ts()}]  BAD UPDATE  {e}"))
    link = connection_status(pv_name, ctx).pipe(
        rx.operators.map(lambda up: f"[{ts()}]  link        {'UP' if up else 'DOWN'}")
    )

    print(f"Resilient monitor on {pv_name} — Ctrl+C to stop")
    print("Try: docker compose stop epics-ioc   (then) docker compose start epics-ioc\n")

    rx.merge(values, errors, link).subscribe(
        on_next=print,
        on_error=lambda e: print(f"FATAL (setup failure): {e}", file=sys.stderr),
        scheduler=scheduler,
    )

    await asyncio.Future()  # run until Ctrl+C


if __name__ == "__main__":
    try:
        asyncio.run(main())
    except KeyboardInterrupt:
        pass

File: RxEpics/cpp/examples/resilient_monitor.cpp
/**
 * Resilient monitor — values, per-update errors, and link state as one merged
 * stream.  Errors and status transitions are messages, not exceptions that stop
 * the process.
 *
 * Mirrors Python's resilient_monitor.py (RxEpics/python/examples/).
 *
 * Run this, then in another shell restart the IOC out from under it:
 *
 *     cd RxEpics/python
 *     docker compose stop epics-ioc
 *     docker compose start epics-ioc
 *
 * The stream survives: link goes DOWN, no traceback, and once the IOC comes back
 * the link goes UP and values resume — PVXS re-arms the subscription on its own,
 * with no client-side action.
 *
 * Usage:
 *   EPICS_PVA_ADDR_LIST=localhost ./resilient_monitor [pv_name]
 *   default: TEST:CALC
 */

#include <chrono>
#include <csignal>
#include <iostream>
#include <string>
#include <vector>

#include <rxcpp/rx.hpp>
#include <rxepics/rxepics.hpp>

static volatile sig_atomic_t g_running = 1;

static long long now_ms() {
    return std::chrono::duration_cast<std::chrono::milliseconds>(
               std::chrono::system_clock::now().time_since_epoch())
        .count();
}

int main(int argc, char* argv[]) {
    const std::string pv = argc > 1 ? argv[1] : "TEST:CALC";

    std::signal(SIGINT, [](int) { g_running = 0; });

    auto& ctx = rxepics::default_context();

    std::cout << "Resilient monitor on " << pv << " — Ctrl+C to stop\n"
              << "Try: (cd RxEpics/python && docker compose stop epics-ioc"
                 " ; docker compose start epics-ioc)\n\n";

    // Three independent streams, tagged and merged into one console view.  Each
    // is a message on the wire, never a terminal notification — a bad update or a
    // dropped link shows up as a line of output, not a crash.  as_dynamic()
    // erases their distinct operator types so they share one vector element type.
    std::vector<rxcpp::observable<std::string>> streams{
        rxepics::monitor_pv<double>(pv, ctx)
            .map([](double v) {
                return "[" + std::to_string(now_ms()) + "]  value       " +
                       std::to_string(v);
            })
            .as_dynamic(),
        rxepics::monitor_errors<double>(pv, ctx)
            .map([](const rxepics::PvUpdateError& e) {
                return "[" + std::to_string(now_ms()) + "]  BAD UPDATE  " +
                       std::string(e.what());
            })
            .as_dynamic(),
        rxepics::connection_status(pv, ctx)
            .map([](bool up) {
                return "[" + std::to_string(now_ms()) + "]  link        " +
                       std::string(up ? "UP" : "DOWN");
            })
            .as_dynamic(),
    };

    auto sub = rxcpp::observable<>::iterate(streams)
        .flat_map([](rxcpp::observable<std::string> s) { return s; })
        .subscribe(
            [](const std::string& line) { std::cout << line << "\n"; },
            [](std::exception_ptr e) {
                try { std::rethrow_exception(e); }
                catch (std::exception& ex) {
                    std::cerr << "FATAL (setup failure): " << ex.what() << "\n";
                }
            });

    while (g_running) std::this_thread::sleep_for(std::chrono::milliseconds(100));
    sub.unsubscribe();
    return 0;
}
